secure/net/src/lib.rs

```diff
@@ -23,55 +23,80 @@ use std::{
 };
 
 pub struct NetworkClient {
-    stream: NetworkStream,
+    server: SocketAddr,
+    stream: Option<NetworkStream>,
 }
 
 impl NetworkClient {
-    pub fn connect(server: SocketAddr) -> Result<Self> {
-        let mut stream = TcpStream::connect(server);
-
-        let mut attempts = 0;
-        let sleeptime = time::Duration::from_millis(100);
-
-        while stream.is_err() && attempts < 5 {
-            thread::sleep(sleeptime);
-            attempts += 1;
-            stream = TcpStream::connect(server);
+    pub fn new(server: SocketAddr) -> Self {
+        Self {
+            server,
+            stream: None,
         }
-        let stream = stream?;
-        stream.set_nodelay(true)?;
-
-        Ok(Self {
-            stream: NetworkStream::new(stream),
-        })
     }
 
     /// Blocking read until able to successfully read an entire message
     pub fn read(&mut self) -> Result<Vec<u8>> {
-        self.stream.read()
+        let stream = self.server()?;
+        let result = stream.read();
+        if result.is_err() {
+            self.stream = None;
+        }
+        result
     }
 
     /// Shutdown the internal network stream
-    pub fn shutdown(&self) -> Result<()> {
-        self.stream.shutdown()
+    pub fn shutdown(&mut self) -> Result<()> {
+        let stream = self
+            .stream
+            .take()
+            .ok_or_else(|| anyhow!("No active stream"))?;
+        stream.shutdown()?;
+        drop(stream);
```

Comment:
why you need explicit drop?

Reply:
Unlikely not necessary but it frees the port for reuse. I wanted to be consistent with the design on the server side.

Reply:
Turns out the explicit take action was strong enough to enforce without needing the drop. Seems shadowing could have resulted in this being the case. The way this is structured eliminates concerns about shadowing etc.